### PR TITLE
nltk dependency parsing tests! 

### PR DESCRIPTION
**Pull request by varmichelle**
- Uses NLTK's Stanford NLP parser to do dependency parsing.
- Generates dependency trees, CoNLL output, and dependency triples for easy analysis
- IMPORTANT: since the jar files the parser needs are too big to commit directly to git, they are downloaded automatically from a website where they're hosted

### CODE REVIEW

**Comment by ltagliaferri** [comment]
I don't see any doctests in the Python file, and it's also not structured either as a function or as OOP. Is there a reason why we went with the NLTK jar files instead of the alternate module? Is having external dependencies the best way to release open-source software?

**Comment by varmichelle** [comment]
Yeah the alternative module would have been preferred, but Assel and I looked into it and the results are noticeably worse than NLTK. Many phrases are mistagged. I discussed with Prof. Cuthbert and this seemed easiest.

External dependencies are annoying but I don't think it's a problem for OSS since Stanford NLP is NLTK's preferred / official / default option. I couldn't find a dependency parser within NLTK that didn't require installing extra stuff. Pretty much every article I saw online uses Stanford's parser.

**Comment by varmichelle** [comment]
Also it looks like there are memory issues:
*******************************************************
***  WARNING!! OUT OF MEMORY! THERE WAS NOT ENOUGH  ***
***  MEMORY TO RUN ALL PARSERS.  EITHER GIVE THE    ***
***  JVM MORE MEMORY, SET THE MAXIMUM SENTENCE      ***
***  LENGTH WITH -maxLength, OR PERHAPS YOU ARE     ***
***  HAPPY TO HAVE THE PARSER FALL BACK TO USING    ***
***  A SIMPLER PARSER FOR VERY LONG SENTENCES.      ***
*******************************************************

Sentence has no parse using PCFG grammar (or no PCFG fallback).  Skipping...
Exception in thread "main" edu.stanford.nlp.parser.common.NoSuchParseException
OSError: Java command failed : ['/usr/bin/java', '-mx1000m', '-cp', 'assets/stanford-parser-3.9.1-models.jar:assets/stanford-parser-3.9.1-models.jar:assets/stanford-parser.jar', 'edu.stanford.nlp.parser.lexparser.LexicalizedParser', '-model', 'edu/stanford/nlp/models/lexparser/englishPCFG.ser.gz', '-sentences', 'newline', '-outputFormat', 'conll2007', '-encoding', 'utf8', '/var/folders/6x/pvsnhlgn7s33yzdq6b859rhc0000gn/T/tmpwy6_o_w6']
	at edu.stanford.nlp.parser.lexparser.LexicalizedParserQuery.getBestParse(LexicalizedParserQuery.java:403)
	at edu.stanford.nlp.parser.lexparser.LexicalizedParserQuery.getBestParse(LexicalizedParserQuery.java:375)
	at edu.stanford.nlp.parser.lexparser.ParseFiles.processResults(ParseFiles.java:272)
	at edu.stanford.nlp.parser.lexparser.ParseFiles.parseFiles(ParseFiles.java:216)
	at edu.stanford.nlp.parser.lexparser.ParseFiles.parseFiles(ParseFiles.java:75)
	at edu.stanford.nlp.parser.lexparser.LexicalizedParser.main(LexicalizedParser.java:1518)

**Comment by varmichelle** [comment]
I haven't written any functions yet, so there aren't any doctests in this file. I figured I should first get this setup pushed so other people on the analysis team can start playing with the parser. We can split tasks from there to write the analysis functions.

File: gender_novels/analysis/dependency_parsing.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+if __name__ == "__main__":
+
+    path_to_jar = "assets/stanford-parser.jar"
+    path_to_models_jar = "assets/stanford-parser-3.9.1-models.jar"
+
+    # The jar files are too big to commit directly, so download them
+    url_to_jar = "http://www.trecento.com/dh_lab/nltk_jar/stanford-parser.jar"
+    url_to_models_jar = "http://www.trecento.com/dh_lab/nltk_jar/stanford-parser-3.9.1-models.jar"
+    urllib.request.urlretrieve(url_to_jar, path_to_jar)
+    urllib.request.urlretrieve(url_to_models_jar, path_to_models_jar)
```
Comment (mscuthbert) [inline gender_novels/analysis/dependency_parsing.py]: make sure to check if the file is already there before downloading -- we don't want a 400MB download everytime something is run, just after it's run the first time.

File: gender_novels/analysis/dependency_parsing.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    dependency_parser = StanfordDependencyParser(path_to_jar, path_to_models_jar)
+
+    novels = Corpus('sample_novels').novels
+
+    novel = novels[0]
+    sentences = sent_tokenize(novel.text)
```
Comment (mscuthbert) [inline gender_novels/analysis/dependency_parsing.py]: If running out of memory -- use one of your teammate's split sentence routines and just tokenize one sentence at a time.  This should be a function to figure out something for one sentence.  Break down the work into the smallest possible functions, writing long scripts to do all at once is 1980s-style programming, so I'm the only one in the lab old enough to be forgiven for doing this.  :-)

File: gender_novels/analysis/dependency_parsing.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        # CoNLL format
+        # print("------CONLL FORMAT------")
+        # print(parse.to_conll(4))
+        # dependency triples of the form ((head word, head tag), rel, (dep word, dep tag))
+        # link defining dependencies: https://nlp.stanford.edu/software/dependencies_manual.pdf
+        print("------DEPENDENCY TRIPLES------")
+        triples = list(parse.triples())
+        for triple in triples:
+            print(triple)
+        print()
```
Comment (mscuthbert) [inline gender_novels/analysis/dependency_parsing.py]: When writing programs, ask yourself how printing things out is going to help others in their analysis -- make small functions or methods that do one thing and then others will be helped.

File: gender_novels/novel.py [hunk 4]
```diff
@@ -145,6 +145,7 @@ def _load_novel_text(self):
             start_novel = text.find('***', end_intro_boilerplate + 5) + 3
             end_novel = text.find('*** END OF THIS PROJECT GUTENBERG EBOOK')
             text = text[start_novel:end_novel]
+            text = text.replace('\n', ' ')
```
Comment (mscuthbert) [inline gender_novels/novel.py]: Looks good to me, but does this break anyone else's work?  I think there are a few tests that this breaks.

File: gender_novels/analysis/dependency_parsing.py (hunk 3 above)
Comment (varmichelle) [reply gender_novels/analysis/dependency_parsing.py]: Sorry, I'm not sure what you mean. This is intended just to print out the dependency triples so the analysis team can work with the information. I haven't gotten much further with analysis so I haven't written more functions. I'm intending to get this pushed first so others can access it and then refactor after I have a better idea of how to analyze.

File: gender_novels/analysis/dependency_parsing.py (hunk 2 above)
Comment (varmichelle) [reply gender_novels/analysis/dependency_parsing.py]: Will do! I've removed the code handling novels for now, but I'll do this when I get to this point.

**Comment by mscuthbert** [comment]
Thanks for the explanations. There are a few things that can be improved but I agree that the big contributions outweigh the small improvements and it’s important to get these capabilities to the group. Merging.

**Pull request merged**